0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,415 @@
+---
+title: Settlement Engine
+draft: 1
+---
+
+# Settlement Engine
+
+## Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Overview
+
+This specification codifies a common interface for a **settlement engine**. Settlement engines are services which perform two primary operations:
+
+1. Send payments: execute outgoing settlements
+2. Receive payments: acknowledge incoming settlements
+
+Counterparties may operate compatible settlement engines to settle their liabilities between one another. Different implementations may utilize different settlement systems or types of settlements, such as:
+
+- Sending money or assets to the counterparty
+- Transferring value on a shared ledger
+- Signing and exchanging payment channel claims
+- Performing a task for the counterparty with a mutually ascribed value
+
+## Usage in Interledger
+
+### Interledger packet flow
+
+In the [Interledger protocol](../0001-interledger-architecture/0001-interledger-architecture.md), a **peer** is a connector who is a counterparty to another connector. Connectors clear and fulfill ILP packets with their peers, which represent conditional IOUs that affect their mutual accounting relationships.
+
+A connector may extend a given peer a limited line of credit, or none at all, depending upon their trustworthiness. As the connector receives incoming ILP PREPARE packets from a peer, forwards them, and returns corresponding ILP FULFILL packets, that peer's liabilities to the connector accrue. If the peer's liabilities exceed the credit limit assigned to it, the connector may reject and decline to forward additional packets.
+
+In order to continue transacting, the peer must settle their liabilities. In most cases, this is accomplished through sending a payment on a settlement system that both peers have agreed to use. The connector should acknowledge the incoming payment&mdash;irrevocably discharging some liability the peer owed to it&mdash;and enabling it to clear and forward subsequent packets from the peer on credit.
+
+Settlement engines provide a standardized mechanism for Interledger connectors to coordinate and reconcile these settlements.
+
+### Accounting
+
+An **account** represents a record of transactions between counterparties, denominated in some fungible asset. Each account may be comprised of multiple account balances, which each represent the net difference between amounts received (credits) and amounts owed (debits) for some subset of their transactions.
```

Comment:
```suggestion
A Connector **account** tracks a debt position between two parties denominated in a single, fungible asset-type.
```

I think its important to call out that Interledger Connectors track debt positions - they don't track the actual asset (this is what the underlying ledgers track). Also, talking about transactions or anything else here seems like an implementation detail that isn't adding value in the context of this RFC.

Reply:
> Also, talking about transactions or anything else here seems like an implementation detail that isn't adding value in the context of this RFC

I disagree. That line is intended to be a precise definition of what an account is, and not stipulate how it's implemented. (It never says that the connector needs to actually track each transaction.)

The next paragraph explains what an Interledger connector is supposed to do and what the accounting system tracks, which is multiple account balances, which are a net amount that include debts and amounts owed.

Reply:
> I disagree. 

I'm not clear about what you disagree with. The current text says, "_An **account** represents a record of transactions between counterparties,_" but existing implementations don't track transactions, so this definition is incorrect at the moment.

Reply:
To me, the use of "represents" indicates it's speaking abstractly about an account, and not expressly saying that's how it's implemented -- is there way to clarify this better?

I don't think simply saying an account is a "debt position" or net position is sufficiently descriptive. A net position _of what_? Even if the connector is not persisting each transaction, each transaction still _happened_; in the ephemeral bookkeeping there's still a credit/debit line for each transaction, e.g. ILP packet fulfilled, settlement received, etc.

Reply:
(Your definition of "account" seems to be my definition of "account balance." I think it's important to differentiate between those two conceptually, but maybe there's more precise wording to use).

Reply:
I think I see your point. Couple of thoughts....

> the use of "represents" indicates it's speaking abstractly about an account, 

I think we shouldn't be abstract here. You're correct that there are a lot of things that _might_ go on underneath an account to arrive at the two balances we care about, but for RFC purposes, these are all out of scope. For example, it's also true that in order to arrive at the list of transactions there is may be a computer system storing the transactions, and auditors that are employed to make sure those transactions aren't tampered with, and on and on. IMHO, these others things that might also represent an account don't matter for this RFC.

That said though, for purposes of this RFC, what _does_ matter? Whatever that answer is, I think we should use that as our definition.

> I don't think simply saying an account is a "debt position" or net position is sufficiently descriptive. A net position of what?

It's a fair point -- though in my mind a "debt position" and a "net position" are two different ideas. A debt-position is a net-position between two parties where the position is denominated in debt (IOUs in this case).

Though I'm not sure that's the absolute thing we should use to describe an account, it does feel closest to, "between two peers, there's an accounts receivable and an accounts payable."

Reply:
Resolving this - I think the latest changes sound good to me.